provider/pkg/provider/kind/kind.go

```diff
@@ -0,0 +1,243 @@
+	clusters := strings.Split(out.String(), "\n")
+	for _, cluster := range clusters {
+		if cluster == clusterName {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// CreateKindCluster orchestrates the creation of a new Kind cluster using Pulumi's local.Command.
+// Validates the input arguments before proceeding with the cluster creation.
+func CreateKindCluster(ctx *pulumi.Context, args *KindClusterArgs, opts ...pulumi.ResourceOption) error {
+	// Validate input arguments.
+	if err := ValidateKindClusterArgs(ctx, *args); err != nil {
+		return err
+	}
+
+	// Run the 'kind create cluster' command.
+	_, err := local.NewCommand(ctx, "createKindCluster", &local.CommandArgs{
+		Create: &schema.Command{
+			Command:        []string{"kind", "create", "cluster", "--name", string(args.ClusterName), "--config", string(args.ConfigFile)},
+			StderrBehavior: schema.CommandBehavior_LOG,
+			StdoutBehavior: schema.CommandBehavior_LOG,
+		},
+	}, opts...)
+
+	if err != nil {
+		ctx.Log.Error("Failed to create Kind cluster", err)
+		return err
+	}
+	ctx.Log.Info("Kind cluster created successfully", nil)
+	return nil
+}
+
+// DeleteKindCluster deletes an existing Kind cluster.
+// Runs the 'kind delete cluster' cli command.
+func DeleteKindCluster(ctx *pulumi.Context, clusterName string, opts ...pulumi.ResourceOption) error {
+	_, err := local.NewCommand(ctx, "deleteKindCluster", &local.CommandArgs{
+		Delete: &schema.Command{
+			Command:        []string{"kind", "delete", "cluster", "--name", clusterName},
+			StderrBehavior: schema.CommandBehavior_LOG,
+			StdoutBehavior: schema.CommandBehavior_LOG,
+		},
+	}, opts...)
+
+	if err != nil {
+		ctx.Log.Error("Failed to delete Kind cluster", err)
+		return err
+	}
+	ctx.Log.Info("Kind cluster deleted successfully", nil)
+	return nil
+}
+
+// UpdateKindCluster updates an existing Kind cluster by deleting and recreating it.
+// A simplistic update strategy which could be optimized in the future.
+func UpdateKindCluster(ctx *pulumi.Context, args *KindClusterArgs) error {
+	// Delete the existing cluster.
+	if err := DeleteKindCluster(ctx, string(args.ClusterName)); err != nil {
+		return err
+	}
+	// Recreate the cluster with the updated configuration.
+	if err := CreateKindCluster(ctx, args); err != nil {
+		return err
+	}
+	return nil
+}
+
+// CreateOrUpdateKindCluster is a high-level function that decides whether to create
+// a new Kind cluster or update an existing one based on the given arguments.
+// It performs validation, existence checks, and then delegates to the appropriate
+// Create or Update method of the KindClusterUtility.
+//
+// Parameters:
+// - ctx: The Pulumi Context object for logging and resource management.
+// - args: The arguments required for Kind cluster creation or update.
+// - preview: A boolean flag for preview mode.
+func CreateOrUpdateKindCluster(ctx *pulumi.Context, args KindClusterArgs, preview bool) error {
```

Comment:
Pulumi will handle the logic of invoking create (if the resource doesn't exists), or update (if it does).

Also: dead code.